cli/pkg/schema/validate.go

```diff
@@ -0,0 +1,102 @@
+package schema
+
+import (
+	"fmt"
+	"slices"
+
+	"github.com/xeipuuv/gojsonschema"
+)
+
+// Perform additional validation checks on the application
+func (a *Application) IsValid() []gojsonschema.ResultError {
+	// Check the names of all resources are unique
+	violations := a.checkNoNameConflicts()
+	violations = append(violations, a.checkNoReservedNames()...)
+
+	return violations
+}
+
+func (a *Application) checkNoNameConflicts() []gojsonschema.ResultError {
+	resourceNames := map[string]string{}
+	violations := []gojsonschema.ResultError{}
+
+	for name := range a.ServiceIntents {
+		if existingType, ok := resourceNames[name]; ok {
+			violations = append(violations, newValidationError(fmt.Sprintf("services.%s", name), fmt.Sprintf("service name %s is already in use by a %s", name, existingType)))
+			continue
+		}
+
+		resourceNames[name] = "service"
+	}
+
+	for name := range a.BucketIntents {
+		if existingType, ok := resourceNames[name]; ok {
+			violations = append(violations, newValidationError(fmt.Sprintf("buckets.%s", name), fmt.Sprintf("bucket name %s is already in use by a %s", name, existingType)))
+			continue
+		}
+		resourceNames[name] = "bucket"
+	}
+
+	for name := range a.EntrypointIntents {
+		if existingType, ok := resourceNames[name]; ok {
+			violations = append(violations, newValidationError(fmt.Sprintf("entrypoints.%s", name), fmt.Sprintf("entrypoint name %s is already in use by a %s", name, existingType)))
+			continue
+		}
+		resourceNames[name] = "entrypoint"
+	}
+
+	for name := range a.DatabaseIntents {
+		if existingType, ok := resourceNames[name]; ok {
+			violations = append(violations, newValidationError(fmt.Sprintf("databases.%s", name), fmt.Sprintf("database name %s is already in use by a %s", name, existingType)))
+			continue
+		}
+		resourceNames[name] = "database"
+	}
+
+	for name := range a.WebsiteIntents {
+		if existingType, ok := resourceNames[name]; ok {
+			violations = append(violations, newValidationError(fmt.Sprintf("websites.%s", name), fmt.Sprintf("website name %s is already in use by a %s", name, existingType)))
+			continue
+		}
+		resourceNames[name] = "website"
+	}
+
+	return violations
+}
+
+func (a *Application) checkNoReservedNames() []gojsonschema.ResultError {
+	violations := []gojsonschema.ResultError{}
+	reservedNames := []string{"backend"}
```

Comment:
I think it's worth us documenting (with a comment for now), why any word is in this list. E.g. in this case due to terraform using backend as a reserved word.

It will give our future selves the info needed to reassess or not touch these values.